internal/components/multi_endpoint.go

```diff
@@ -0,0 +1,95 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package components
+
+import (
+	"fmt"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/open-telemetry/opentelemetry-operator/internal/naming"
+)
+
+var _ ComponentPortParser = &MultiPortReceiver{}
+
+// MultiProtocolEndpointConfig represents the minimal struct for a given YAML configuration input containing a map to
+// a struct with either endpoint or listen_address.
+type MultiProtocolEndpointConfig struct {
+	Protocols map[string]*SingleEndpointConfig `json:"protocols"`
+}
+
+// MultiPortOption allows the setting of options for a MultiPortReceiver.
+type MultiPortOption func(parser *MultiPortReceiver)
+
+// MultiPortReceiver is a special parser for components with endpoints for each protocol.
+type MultiPortReceiver struct {
+	name string
+
+	portMappings map[string]*corev1.ServicePort
+}
+
+func (m *MultiPortReceiver) Ports(logger logr.Logger, config interface{}) ([]corev1.ServicePort, error) {
+	multiProtoEndpointCfg := &MultiProtocolEndpointConfig{}
+	if err := LoadMap[*MultiProtocolEndpointConfig](config, multiProtoEndpointCfg); err != nil {
+		return nil, err
+	}
+	var ports []corev1.ServicePort
+	for protocol, ec := range multiProtoEndpointCfg.Protocols {
+		if defaultSvc, ok := m.portMappings[protocol]; ok {
+			port := defaultSvc.Port
+			if ec != nil {
+				port = ec.GetPortNumOrDefault(logger, port)
+				defaultSvc.Name = naming.PortName(fmt.Sprintf("%s-%s", m.name, protocol), port)
+			}
+			ports = append(ports, ConstructServicePort(defaultSvc, port))
+		} else {
+			return nil, fmt.Errorf("unknown protocol set: %s", protocol)
+		}
+	}
+	return ports, nil
+}
+
+func (m *MultiPortReceiver) ParserType() string {
+	return ComponentType(m.name)
+}
+
+func (m *MultiPortReceiver) ParserName() string {
+	return fmt.Sprintf("__%s", m.name)
```

Comment:
Why the underscores here?

Reply:
All the parsers had `__` as a prefix: https://github.com/open-telemetry/opentelemetry-operator/blob/59a7c258dcf3517870a72762e31e8c118b036f59/internal/manifests/collector/parser/receiver/receiver_aws-xray.go#L23

Reply:
yeah this is for compatibility, all of that is internal so it doesn't __really__ matter but I thought i would keep it consistent to not break other tests in the future.